src/clients/docs_generate.zig

```diff
@@ -0,0 +1,265 @@
+//     var cp = std.child_process.ChildProcess.init("docker", &[_][]const u8{
+//         "run",
+//         language,
+//         },
+//                                                  );
+// }
+
+// pub fn validate_docs(language: Docs) {
+//     try run_in_docker(
+//         language.test_linux_docker_image,
+//         language.install,
+//     );
+//     try run_in_docker(
+//         language.test_linux_docker_image,
+//         language.developer_setup_bash_commands,
+//     );
+// }
+
+const MarkdownWriter = struct {
+    buf: *std.ArrayList(u8),
+    writer: std.ArrayList(u8).Writer,
+
+    fn init(buf: *std.ArrayList(u8)) MarkdownWriter {
+        return MarkdownWriter{ .buf = buf, .writer = buf.writer() };
+    }
+
+    fn header(mw: *MarkdownWriter, comptime n: i8, content: []const u8) void {
+        mw.print(("#" ** n) ++ " {s}\n\n", .{content});
+    }
+
+    fn paragraph(mw: *MarkdownWriter, content: []const u8) void {
+        // Don't print empty lines.
+        if (content.len == 0) {
+            return;
+        }
+        mw.print("{s}\n\n", .{content});
+    }
+
+    fn code(mw: *MarkdownWriter, language: []const u8, content: []const u8) void {
+        // Don't print empty lines.
+        if (content.len == 0) {
+            return;
+        }
+        mw.print("```{s}\n{s}\n```\n\n", .{ language, content });
+    }
+
+    fn commands(mw: *MarkdownWriter, content: []const u8) void {
+        mw.print("```console\n", .{});
+        var splits = std.mem.split(u8, content, "\n");
+        while (splits.next()) |chunk| {
+            mw.print("$ {s}\n", .{chunk});
+        }
+
+        mw.print("```\n\n", .{});
+    }
+
+    fn print(mw: *MarkdownWriter, comptime fmt: []const u8, args: anytype) void {
+        try mw.writer.print(fmt, args) catch unreachable;
+    }
+
+    fn reset(mw: *MarkdownWriter) void {
+        mw.buf.clearRetainingCapacity();
+    }
+
+    fn diffOnDisk(mw: *MarkdownWriter, filename: []const u8) !bool {
+        const file = try std.fs.cwd().createFile(filename, .{ .read = true, .truncate = false });
+        const fSize = (try file.stat()).size;
+        if (fSize != mw.buf.items.len) {
+            return true;
+        }
+
+        var h = std.crypto.hash.sha2.Sha256.init(.{});
+        var buf = std.mem.zeroes([4096]u8);
+        var cursor: usize = 0;
+        while (cursor < fSize) {
+            var maxCanRead = if (fSize - cursor > 4096) 4096 else fSize - cursor;
+            var read = try file.read(buf[0..maxCanRead]);
+            h.update(buf[0..read]);
+            cursor += read;
+        }
+
+        var newH = std.crypto.hash.sha2.Sha256.init(.{});
+        newH.update(mw.buf.items);
+        var newHash: [std.crypto.hash.sha2.Sha256.digest_length]u8 = undefined;
+        newH.final(newHash[0..]);
+        var existingHash: [std.crypto.hash.sha2.Sha256.digest_length]u8 = undefined;
+        h.final(existingHash[0..]);
```

Comment:
Why hash instead of just comparing the bytes directly?

Reply:
haha, great point. dropping.